login/service.go

```diff
@@ -201,32 +204,71 @@ func (keycloak *KeycloakOAuthProvider) Exchange(ctx context.Context, code string
 
 // ExchangeRefreshToken exchanges refreshToken for OauthToken
 func (keycloak *KeycloakOAuthProvider) ExchangeRefreshToken(ctx context.Context, refreshToken string, endpoint string, serviceConfig LoginServiceConfiguration) (*token.TokenSet, error) {
-	identity, err := LoadContextIdentityAndUser(ctx, keycloak.DB)
+
+	// Load identity for the refresh token
+	var identity *account.Identity
+	claims, err := keycloak.TokenManager.ParseTokenWithMapClaims(ctx, refreshToken)
+	if err != nil {
+		return nil, autherrors.NewUnauthorizedError(err.Error())
+	}
+	sub := claims["sub"]
+	if sub == nil {
+		return nil, autherrors.NewUnauthorizedError("missing 'sub' claim in the refresh token")
+	}
+	identityID, err := uuid.FromString(fmt.Sprintf("%s", sub))
+	if err != nil {
+		return nil, autherrors.NewUnauthorizedError(err.Error())
+	}
+	err = application.Transactional(keycloak.DB, func(appl application.Application) error {
+		identity, err = appl.Identities().LoadWithUser(ctx, identityID)
+		return err
+	})
+	if err != nil {
+		// That's OK if we didn't find the identity if the token was issued for an API client
+		// Just log it and proceed.
+		log.Warn(ctx, map[string]interface{}{
+			"err": err,
+		}, "failed to load identity when refreshing token; it's OK if the token was issued for an API client")
+	}
 	if identity != nil && identity.User.Deprovisioned {
 		log.Warn(ctx, map[string]interface{}{
 			"identity_id": identity.ID,
 			"user_name":   identity.Username,
 		}, "deprovisioned user tried to refresh token")
 		return nil, autherrors.NewUnauthorizedError("unauthorized access")
 	}
+
+	// Refresh token in Keycloak
+	tokeSet, err := keycloak.keycloakTokenService.RefreshToken(ctx, endpoint, serviceConfig.GetKeycloakClientID(), serviceConfig.GetKeycloakSecret(), refreshToken)
```

Comment:
What do you think of the following workflow -

On err,
1. Check if `err` is `not because of` incorrect refresh token / creds ( network problem, etc )
2. and generate tokens locally, and move on. 

That way everytime kc is down, the automatic fallback is local token generation. :)

Reply:
Currently KC relays on login sessions. I would keep it for now. We are going to implement token retrospective mechanism in Auth (as part of AuthZ) later too.